// webpack-bundle-analyzer/lib/BundleAnalyzerPlugin.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var fs = require('fs');
var path = require('path');
var mkdir = require('mkdirp');

var _require = require('chalk'),
    bold = _require.bold;

var viewer = require('./viewer');

var BundleAnalyzerPlugin = function () {
  function BundleAnalyzerPlugin(opts) {
    _classCallCheck(this, BundleAnalyzerPlugin);

    this.opts = _extends({
      analyzerMode: 'server',
      analyzerPort: 8888,
      reportFilename: 'report.html',
      openAnalyzer: true,
      generateStatsFile: false,
      statsFilename: 'stats.json',
      statsOptions: null,
      // deprecated
      startAnalyzer: true
    }, opts);
  }

  BundleAnalyzerPlugin.prototype.apply = function apply(compiler) {
    var _this = this;

    this.compiler = compiler;

    compiler.plugin('done', function (stats) {
      stats = stats.toJson(_this.opts.statsOptions);

      var actions = [];

      if (_this.opts.generateStatsFile) {
        actions.push(function () {
          return _this.generateStatsFile(stats);
        });
      }

      // Handling deprecated `startAnalyzer` flag
      if (_this.opts.analyzerMode === 'server' && !_this.opts.startAnalyzer) {
        _this.opts.analyzerMode = 'disabled';
      }

      if (_this.opts.analyzerMode === 'server') {
        actions.push(function () {
          return _this.startAnalyzerServer(stats);
        });
      } else if (_this.opts.analyzerMode === 'static') {
        actions.push(function () {
          return _this.generateStaticReport(stats);
        });
      }

      if (actions.length) {
        // Making analyzer logs to be after all webpack logs in the console
        setTimeout(function () {
          console.log('');
          actions.forEach(function (action) {
            return action();
          });
        }, 200);
      }
    });
  };

  BundleAnalyzerPlugin.prototype.generateStatsFile = function generateStatsFile(stats) {
    var statsFilepath = this.opts.statsFilename;

    if (!path.isAbsolute(statsFilepath)) {
      statsFilepath = path.resolve(this.compiler.outputPath, statsFilepath);
    }

    mkdir.sync(path.dirname(statsFilepath));

    fs.writeFileSync(statsFilepath, JSON.stringify(stats, null, 2));

    console.log(bold('Webpack Bundle Analyzer') + ' saved stats file to ' + bold(statsFilepath));
  };

  BundleAnalyzerPlugin.prototype.startAnalyzerServer = function startAnalyzerServer(stats) {
    viewer.startServer(stats, {
      openBrowser: this.opts.openAnalyzer,
      port: this.opts.analyzerPort,
      bundleDir: this.compiler.outputPath
    });
  };

  BundleAnalyzerPlugin.prototype.generateStaticReport = function generateStaticReport(stats) {
    viewer.generateReport(stats, {
      openBrowser: this.opts.openAnalyzer,
      reportFilename: this.opts.reportFilename,
      bundleDir: this.compiler.outputPath
    });
  };

  return BundleAnalyzerPlugin;
}();

module.exports = BundleAnalyzerPlugin;
